Add: NewGRF Badges feature

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Authors of vehicle sets like to add symbols to preview sprites in purchase lists. However across sets they don't generally agree on what the symbols should be, how large they are, and where they are placed.

From a user perspective this can end up looking messy and cluttered.

As they are part of the preview sprite, there is no opportunity to filter on them, and they cannot be hidden.

![image](https://github.com/user-attachments/assets/f5fcc7a5-dbb3-4ced-bb3b-3630e404d290)


<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Enter NewGRF 'badges'.

A badge is a string-based label that can be defined by any NewGRF, and applied to most NewGRF features by way of a label 'translation' table. Badges can be assigned even if they are not defined, in this case they are attached but do not have any visual appearance.

Badges are grouped by 'class', which is defined by the first part of the string label up to a `/` separator.

I settled on an non-conventional approach of using strings as it reduces problems of clashes (as badges are global), and it allows authors to drill down into whatever detail they like. This differs from existing label schemes that use 4 byte values (normally treated as 4 alphanumeric characters), which would require compromises.

Badges appear in purchase lists, grouped into columns. An feature item can have multiple badges assigned, if any of them are in the same class only the first badge of that class is displayed in the main purchase list, although all badge descriptions are displayed in the extra-text part.

![image](https://github.com/user-attachments/assets/3f7ece3d-6e19-4182-9742-3489b73cfffb)
![image](https://github.com/user-attachments/assets/0404f4fb-93ff-4452-8d7c-fcdec900163b)

Display of badges can be customised by players. There is a rudimentary way to do this for vehicle purchase lists, but other features currently need to be configured in a configuration file.

To make NewGRF badges useful for more just displaying symbols and filter the purchase list, it is also possible to query for them in var action chains when resolving sprites or callbacks. This allows badges (defined or undefined) to be used as generic metadata by NewGRFs.

The intention is to also have a set of default badges for authors to use. As that is currently a NewGRF generated by grf-py that is separate from this for now.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

Needs a better way to get to user configuration which is currently only available for engines, even though badges are available on other features.

And maybe a better way to filter by badge other than by typing its name. A drop down list similar to cargo types, perhaps, but I don't know where/how to fit that in.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Add: NewGRF Badges.

rubidium42 commented on src/autoreplace_gui.cpp:
```diff
@@ -464,7 +468,7 @@ class ReplaceVehicleWindow : public Window {
 
 				/* Do the actual drawing */
-				DrawEngineList((VehicleType)this->window_number, r, this->engines[side], *this->vscroll[side], this->sel_engine[side], side == 0, this->sel_group);
+				DrawEngineList((VehicleType)this->window_number, r, this->engines[side], *this->vscroll[side], this->sel_engine[side], side == 0, this->sel_group, this->badge_classes);
```
Maybe `static_cast<VehicleType>` as you're touching the line?

rubidium42 commented on src/build_vehicle_gui.cpp:
```diff
@@ -990,6 +993,11 @@ int DrawVehiclePurchaseInfo(int left, int right, int y, EngineID engine_number,
 	return y;
 }
 
+static void DrawEngineBadgeColumn(Rect r, int column_group, const GUIBadgeClasses &badge_classes, const Engine *e, PaletteID remap)
```
Do you need a copy of `Rect` here, or is `const Rect &` enough?

rubidium42 commented on src/newgrf_badge.cpp:
```diff
@@ -0,0 +1,516 @@
+{
+	BadgeResolverObject object(badge, feature, introduction_date);
+	const SpriteGroup *group = object.Resolve();
+	if (group == nullptr) return {0, PAL_NONE};
```
A named constant for the `0` here?

PeterN replied:
There isn't one. We use `{0, PAL_NONE}` all over the place already to indicate that nothing should be drawn. Sprite 0 is actually the regular mouse cursor.

rubidium42 commented on src/lang/english.txt:
```diff
@@ -5901,3 +5903,9 @@ STR_PLANE                                                       :{BLACK}{PLANE}
+STR_BADGE_NAME_LIST                                             :{STRING}: {GOLD}{RAW_STRING}
+STR_BADGE_CONFIG_RESET                                          :Reset
+STR_BADGE_CONFIG_PREVIEW                                        :Preview Image
+STR_BADGE_CONFIG_NAME                                           :Name
```
Except `STR_BADGE_NAME_LIST` none of these are used in commit that adds these.

rubidium42 commented on src/dropdown.cpp:
```diff
@@ -68,6 +68,8 @@ static WindowDesc _dropdown_desc(
 	_nested_dropdown_menu_widgets
 );
 
+Point _dropdown_hit;
```
These drop down changes don't seem to be used in the commit that adds them,

rubidium42 commented on src/newgrf_badge.cpp:
```diff
@@ -389,6 +390,42 @@ uint GUIBadgeClasses::GetTotalColumnsWidth() const
+
+	/* Pre-build list of badges that match by string. */
+	for (const auto &badge : _badges.specs) {
+		if (badge.name == 0) continue;
```
`STR_NULL` instead of 0?

rubidium42 commented on src/build_vehicle_gui.cpp:
```diff
@@ -1215,6 +1219,7 @@ struct BuildVehicleWindow : Window {
 	} filter;                                   ///< Filter to apply.
 	bool descending_sort_order;                 ///< Sort direction, @see _engine_sort_direction
 	uint8_t sort_criteria;                         ///< Current sort criterium.
+
```
Spurious newline (in commit)?

rubidium42 commented on src/build_vehicle_gui.cpp:
```diff
@@ -1665,6 +1670,12 @@ struct BuildVehicleWindow : Window {
+	DropDownList BuildBadgeConfigurationList() const
+	{
+		static const auto separators = { STR_BADGE_CONFIG_PREVIEW, STR_BADGE_CONFIG_NAME };
+		return ::BuildBadgeClassConfigurationList(static_cast<GrfSpecFeature>(GSF_TRAINS + this->vehicle_type), 3, separators);
```
Please use a named constant for the number of columns, as I'm seeing this 3 being used later in the class. Similarly there's a 2 that I suspect to be 'number of columns - 1'.

rubidium42 reviewed (commented):
I've mostly skimmed through the changes commit-by-commit. I have to go through them in more detail at a later point.

rubidium42 commented on src/build_vehicle_gui.cpp:
```diff
@@ -1881,6 +1961,40 @@ struct BuildVehicleWindow : Window {
+				} else {
+					/* Toggle */
+					BadgeClassToggleVisibility(feature, BadgeClassID(index));
+				}
```
This looks weird to me. Comparing the difference in X and Y location to the width? So is it intentional that there are diagonal bands through the widget with differing behaviour?

If that's the case, then maybe a small comment that such behaviour is what you're looking for. Though, aren't there width/height combinations where you can't reach the toggle functionality with this logic?

PeterN replied:
Yeah, this is a complete bodge which abuses Point. I kinda came up with it as a hack and it stuck, whoops.

Ideally we'd have the dropdown element itself be able to intercept the dropdown click response and add another parameter or something, and not using globals.

rubidium42 commented on src/newgrf.cpp:
```diff
@@ -4826,6 +4967,50 @@ static ChangeInfoResult IgnoreRoadStopProperty(uint prop, ByteReader &buf)
 
+static ChangeInfoResult BadgeChangeInfo(uint first, uint last, int prop, ByteReader &buf)
+{
+	static const int NUM_BADGES_PER_GRF = 16384;
```
Isn't this a few times bigger than the number of distinct short names that can be added through action13? Would it make sense to allow more badges than potential (unique) short names?

PeterN replied:
Actually the short name property isn't needed at all and I shall get rid of it. It's a left-over from an earlier iteration.

Names are set via a non-generic Action 4, same as vehicles, so there is no limit to the names here (and it doesn't eat into the limited space of 0xD0/0xDC strings.)

rubidium42 commented on src/newgrf.cpp:
```diff
@@ -6180,6 +6371,56 @@ static void RoadStopMapSpriteGroup(ByteReader &buf, uint8_t idcount)
+	std::vector<uint16_t> local_ids;
+	local_ids.reserve(idcount);
+	for (uint i = 0; i < idcount; i++) {
+		local_ids.push_back(buf.ReadExtendedByte());
```
Weren't you more-or-less abolishing extended bytes in your GRFv9 PR? Or did I misunderstood that change?

PeterN replied:
It did but this isn't that PR. I used ExtendedByte here to maintain parity -- for all features the action 3 layout is the same.

rubidium42 commented on src/newgrf_badge.cpp:
```diff
@@ -0,0 +1,564 @@
+		/* Is badge copyable? */
+		Badge *badge = GetBadge(index);
+		if (badge == nullptr) continue;
+		if ((badge->flags & BadgeFlags::Copy) == BadgeFlags::None) continue;
```
Didn't you introduce `HasFlag` for this? Or rather `!HasFlag` in this situation.

rubidium42 commented on src/newgrf_badge.cpp:
```diff
@@ -0,0 +1,564 @@
+	}
+}
+
+static constexpr uint MAX_BADGE_HEIGHT = 12; ///< Maximmal height of a badge sprite.
```
```suggestion
static constexpr uint MAX_BADGE_HEIGHT = 12; ///< Maximal height of a badge sprite.
```

rubidium42 commented on src/newgrf_badge.cpp:
```diff
@@ -0,0 +1,564 @@
+}
+
+/**
+ * Test if one of the lists badges matches the filtered badge list.
```
```suggestion
 * Test if one of the list's badges matches the filtered badge list.
```
Or maybe just 'of the given badges'?

PeterN committed:
Codechange: Make non-generic GRF string ids unique per feature.

This prevents engine name strings conflicting with generic strings if more than 53,248 engines are defined.

PeterN committed:
Codechange: Allow iterating a BaseBitSet.

rubidium42 commented on src/newgrf.cpp:
```diff
@@ -6211,6 +6397,56 @@ static void RoadStopMapSpriteGroup(ByteReader &buf, uint8_t idcount)
+		local_ids.push_back(buf.ReadExtendedByte());
+	}
+
+	uint8_t cidcount = buf.ReadByte();
+	for (uint c = 0; c < cidcount; c++) {
+		uint8_t ctype = buf.ReadByte();
+		uint16_t groupid = buf.ReadWord();
+		if (!IsValidGroupID(groupid, "BadgeMapSpriteGroup")) continue;
+
+		if (ctype >= GSF_END) continue;
+
+		for (const auto &local_id : local_ids) {
+			auto found = _cur.grffile->badge_map.find(local_id);
+			if (found == std::end(_cur.grffile->badge_map)) {
+				GrfMsg(1, "BadgeMapSpriteGroup: Badge {} undefined, skipping", local_id);
+				continue;
+			}
+
+			auto &badge = *GetBadge(found->second);
+			badge.grf_prop.SetSpriteGroup(ctype, _cur.spritegroups[groupid]);
+		}
+	}
```
What does the `c` and c in `cidcount` stand for? Smells like 'cargo', but it's compared against `GrfSpecFeature`.

PeterN replied:
"Yes", it's common code pattern used but duplicated by all features. `cid` and `cidcount` are the names in the Action 3 spec, although of course they derive from cargo types which was the usage originally.

These names are still used for (most) other features around here.
